lib/knapsack_pro/runners/queue/rspec_runner.rb

```diff
@@ -3,224 +3,287 @@
+            queue_runner.run
           end
         end
 
-        def self.ensure_spec_opts_have_knapsack_pro_formatters
-          return unless ENV['SPEC_OPTS']
+        def initialize(adapter_class, function_core, args, stream_error, stream_out)
+          super(adapter_class)
+          @adapter_class = adapter_class
+          @function_core = function_core
+          @cli_args = prepare_cli_args(args)
+          @stream_error = stream_error
+          @stream_out = stream_out
+          @node_assigned_test_file_paths = []
+          @rspec_runner = nil # RSpec::Core::Runner is lazy initialized
+        end
 
-          if [
-              ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s),
-              ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::TimeTracker.to_s),
-          ].all?
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L85
+        #
+        # @return [Fixnum] exit status code.
+        #   0 if all specs passed,
+        #   or the configured failure exit code (1 by default) if specs failed.
+        def run
+          pre_run_setup
+
+          if @rspec_runner.knapsack__wants_to_quit?
+            exit_code = @rspec_runner.knapsack__exit_early
+            Kernel.exit(exit_code)
             return
           end
 
-          unless ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s)
-            ENV['SPEC_OPTS'] = "#{ENV['SPEC_OPTS']} --format #{KnapsackPro::Formatters::RSpecQueueSummaryFormatter}"
+          begin
+            exit_code = @rspec_runner.knapsack__run_specs(self)
+          rescue KnapsackPro::Runners::Queue::BaseRunner::TerminationError
+            @function_core.set_error_exit_code(@rspec_runner.knapsack__error_exit_code)
+            raise
+          rescue Exception => exception
+            KnapsackPro.logger.error("An unexpected exception happened. RSpec cannot handle it. The exception: #{exception.inspect}")
+            @function_core.log_exit_summary(@node_assigned_test_file_paths)
+            @function_core.set_error_exit_code(@rspec_runner.knapsack__error_exit_code)
+            raise
           end
 
-          unless ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::TimeTracker.to_s)
-            ENV['SPEC_OPTS'] = "#{ENV['SPEC_OPTS']} --format #{KnapsackPro::Formatters::TimeTracker}"
+          post_run_tasks(exit_code)
+        end
+
+        def with_batched_tests_from_queue
+          can_initialize_queue = true
+
+          loop do
+            handle_signal!
+            test_file_paths = pull_tests_from_queue(can_initialize_queue: can_initialize_queue)
+            can_initialize_queue = false
+
+            break if test_file_paths.empty?
+
+            with_queue_hooks(test_file_paths) do |wrapped_test_file_paths|
+              yield wrapped_test_file_paths
+            end
```

Comment:
I'd stay with a different name to avoid overlapping the existing `test_file_paths` variable.

`wrapped_test_file_paths` are coming from inside of `with_queue_hooks` and are yielded out of `with_batched_tests_from_queue`.

Reply:
But they are not "wrapped", what's passed to the method is `yield`ed, so they are the same `test_file_paths`. That's why I'd use the same name (the shadowing should not create problems).